### PR TITLE
Conseil 279 aggregation function support

### PR DESCRIPTION
**Pull request by piotrkosecki**
Added support for aggregation function described in issue #279, refactored query validation and added new function for the reading from cache without updating current counts.

### CODE REVIEW

File: src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        output = None,
+        aggregation = Some(Aggregation(field = "valid"))
+      )
+
+      val result = query.validate("test", tpdo)
+
+      result.futureValue.right.get shouldBe Query(fields = List("valid"), aggregation = Some(Aggregation(field = "valid")))
+    }
+
+    "return error with incorrect aggregation field which exists in DB" in {
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala]: Can the test description make it apparent that the issue is that we're trying to aggregate on a non-aggregating field type?

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 2]
```diff
@@ -854,14 +854,13 @@ class TezosDatabaseOperationsTest
...
       val columns = List()
-
       val populateAndTest = for {
         _ <- Tables.Blocks ++= blocksTmp
         found <- sut.selectWithPredicates(Tables.Blocks.baseTableRow.tableName, columns, List.empty, List.empty, 3)
       } yield found
 
       val result = dbHandler.run(populateAndTest.transactionally).futureValue
-      result shouldBe List(
+      result should contain allElementsOf List(
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: or `contain theSameElementsAs` ?

There's a subtle difference

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 3]
```diff
@@ -1598,6 +1597,102 @@ class TezosDatabaseOperationsTest
...
+        FeesRow(0, 3, 4, new Timestamp(2), "kind")
+      )
+
+      val aggregate = Some(
+        Aggregation("medium", AggregationType.count, None)
+      )
+
+      val populateAndTest = for {
+        _ <- Tables.Fees ++= feesTmp
+        found <- sut.selectWithPredicates(Tables.Fees.baseTableRow.tableName, List("low", "medium", "high"), List.empty, List.empty, 3, aggregate)
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: Using named arguments in the call can help out to figure out what these values represents, e.g.

```scala
sut.selectWithPredicates(
  table = ...,
  columns = ...,
  predicates = ...,
  ordering = ...
  ...
)
```

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 4]
```diff
@@ -26,6 +29,69 @@ object DataTypes {
   /** Max value of limit parameter */
   val maxLimitValue: Int = 100000
 
+  /** Helper method for finding invalid fields in query */
+  private def findInvalidQueryFields(query: Query, entity: String, tezosPlatformDiscovery: TezosPlatformDiscoveryOperations)
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: yes, I've found a way, maybe it is not as clean, but it does not reuse so much code, made change

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 5]
```diff
@@ -26,6 +29,69 @@ object DataTypes {
...
+    (implicit ec: ExecutionContext): Future[List[InvalidQueryField]] = {
+    query
+      .fields
+      .traverse(field => tezosPlatformDiscovery.areFieldsValid(entity, Set(field)).map(_ -> field))
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: I used wrong method, which I removed as it is not used in other places - I don't need one which takes `Set` as it does not return info which fields are invalid.

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 6]
```diff
@@ -34,26 +100,17 @@ object DataTypes {
...
-  class OperationTypeRef extends TypeReference[OperationType.type]
-
-  /** Class required for OutputType enum serialization */
-  class OutputTypeRef extends TypeReference[OutputType.type]
-
   /** Class representing predicate */
   case class Predicate(
     field: String,
-    @JsonScalaEnumeration(classOf[OperationTypeRef]) operation: OperationType,
+    operation: OperationType,
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: found one more place and removed those annotations 😄

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 7]
```diff
@@ -26,6 +29,69 @@ object DataTypes {
...
+      .map(_.collect { case (false, fieldName) => InvalidOrderByField(fieldName) })
+
+  }
+
+  /** Helper method for finding invalid fields in aggregation */
+  private def findInvalidAggregationFields(query: Query, entity: String, tezosPlatformDiscovery: TezosPlatformDiscoveryOperations)
+    (implicit ec: ExecutionContext): Future[List[InvalidAggregationField]] = {
+    query
+      .aggregation
+      .toList
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: changed implementation, this is no longer valid

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 8]
```diff
@@ -26,6 +29,69 @@ object DataTypes {
...
+              .find(_.name == aggregation.field)
+              .map(attribute => canBeAggregated(attribute.dataType) -> aggregation.field)
+          }
+        }
+      }
+      .map(_.flatten.collect { case (false, fieldName) => InvalidAggregationFieldForType(fieldName) }.toList)
+  }
+
+  /** Method checks if type can be aggregated */
+  private def canBeAggregated(tpe: DataType): Boolean = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: ok, changed

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 9]
```diff
@@ -64,56 +121,67 @@ object DataTypes {
...
     predicates: List[Predicate] = List.empty,
     orderBy: List[QueryOrdering] = List.empty,
     limit: Int = defaultLimitValue,
-    output: OutputType = OutputType.json
+    output: OutputType = OutputType.json,
+    aggregation: Option[Aggregation] = None
   )
 
+  /** Class representing predicate used in aggregation */
+  case class AggregationPredicate(
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: no we can't from the reasons you mentioned in the second comment - we could but I don't think it's worth it, to perform additional validation for something what should be verified at the parsing json

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataJsonSchemas.scala [hunk 10]
```diff
@@ -32,6 +32,17 @@ trait DataJsonSchemas extends algebra.JsonSchemas with generic.JsonSchemas {
   implicit def queryOutputSchema: JsonSchema[OutputType.Value] =
     enumeration(OutputType.values.toSeq)(_.toString)
 
+  /** Query aggregation schema */
+  implicit def queryAggregationSchema: JsonSchema[Aggregation] =
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataJsonSchemas.scala]: changed to lazy vals

File: src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala [hunk 11]
```diff
@@ -404,7 +404,8 @@ object ApiOperations extends DataOperations with MetadataOperations {
         query.fields,
         sanitizePredicates(query.predicates),
         query.orderBy,
-        Math.min(query.limit, DataTypes.maxLimitValue)
+        Math.min(query.limit, DataTypes.maxLimitValue),
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 12]
```diff
@@ -250,6 +250,20 @@ class TezosPlatformDiscoveryOperations(
...
   }
 
+  /**
+    * Extracts attributes in the DB for the given table name without updating counts
+    *
+    * @param  tableName name of the table from which we extract attributes
+    * @return list of attributes as a Future
+    */
+  def getTableAttributesWithoutCounts(tableName: String): Future[Option[List[Attribute]]] = {
+    attributesCache.read.map { entitiesMap =>
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: extracted the type aliases for the things that are inside of the MVars

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala [hunk 13]
```diff
@@ -250,6 +250,20 @@ class TezosPlatformDiscoveryOperations(
     }.unsafeToFuture()
   }
 
+  /**
+    * Extracts attributes in the DB for the given table name without updating counts
+    *
+    * @param  tableName name of the table from which we extract attributes
+    * @return list of attributes as a Future
+    */
+  def getTableAttributesWithoutCounts(tableName: String): Future[Option[List[Attribute]]] = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: changed to something more descriptive, `getTableAttributesWithoutUpdatingCache` is ok?

File: src/main/scala/tech/cryptonomic/conseil/util/DatabaseUtil.scala [hunk 14]
```diff
@@ -91,6 +92,21 @@ object DatabaseUtil {
...
       }
+
+      /** Method for adding group by to existing SQLAction
+        *
+        * @param aggregation parameter containing info about field which has to be aggregated
+        * @param columns     parameter containing columns which chich are being used in query
+        * @return new SQLActionBuilder containing limit statement
+        */
+      def addGroupBy(aggregation: Option[Aggregation], columns: List[String]): SQLActionBuilder = {
+        val columnsWithoutAggregation = columns.filterNot(aggregation.map(_.field).contains(_))
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/util/DatabaseUtil.scala]: changed to `columns.filterNot(col => aggregation.exists(_.field == col))` as you proposed

File: src/main/scala/tech/cryptonomic/conseil/util/DatabaseUtil.scala [hunk 15]
```diff
@@ -111,10 +127,15 @@ object DatabaseUtil {
...
       * @param table   table on which query will be executed
       * @param columns columns which are selected from teh table
+      * @param aggregation parameter containing info about field which has to be aggregated
       * @return SQLAction with basic query
       */
-    def makeQuery(table: String, columns: List[String]): SQLActionBuilder = {
-      val cols = if (columns.isEmpty) "*" else columns.mkString(",")
+    def makeQuery(table: String, columns: List[String], aggregation: Option[Aggregation]): SQLActionBuilder = {
+      val aggr = columns.foldLeft(List.empty[String]) {
+        case (acc, column) if aggregation.map(_.field).contains(column) => mapAggregationToSQL(aggregation.get.function, column) :: acc
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/util/DatabaseUtil.scala]: changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala (hunk 1 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala]: changed to be more descriptive

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: you are absolutely right! my bad, changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 16]
```diff
@@ -14,18 +15,57 @@ import scala.concurrent.{ExecutionContext, Future}
...
+  lazy val canBeAggregated: DataType => Boolean = Set(DataType.Decimal, DataType.Int, DataType.LargeInt, DataType.DateTime)
   /** Default value of limit parameter */
   val defaultLimitValue: Int = 10000
   /** Max value of limit parameter */
   val maxLimitValue: Int = 100000
 
+  /** Helper method for finding fields used in query that don't exist in the database */
+  private def findNonExistingFields(query: Query, entity: String, tezosPlatformDiscovery: TezosPlatformDiscoveryOperations)
+    (implicit ec: ExecutionContext): Future[List[QueryValidationError]] = {
+    val fields = query.fields.map("query" -> _) ::: query.predicates.map("predicate" -> _.field) :::
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: No need to change, just sharing ideas.
I find this is a perfect place to use scala Symbol, at least to me it reads better, thought technically there's no difference.

```scala
query.fields.map('query -> _)  :::
  query.predicates.map('predicate -> _.field) :::
  query.orderBy.map('orderBy -> _.field) :::
  [...]

And the same Symbol used to pattern match
```

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: I find it's awesome now

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: Luvely

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala]: awesome

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 16 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: good point, changed

**Pull request merged**